mobgap/sl/_stride_length.py

```diff
@@ -0,0 +1,296 @@
+        # 2. High-pass filtering --> lower cut-off: 0.1 Hz, filter design: Butterworth IIR, filter order: 4
+        vacc = newAcc["acc_x"]  # TODO: check that acceleration is passed to the calculate method in m/s^2
+        fc = 0.1  # cut-off frequency
+        [df_, cf_] = butter(4, fc / (sampling_rate_hz / 2), "high")  # design high-pass filter
+        vacc_high = lfilter(df_, cf_, vacc)  # filtering
+
+        # 3. Integration of vertical acceleration --> vertical speed
+        # 4. Drift removal (high-pass filtering) --> lower cut-off: 1 Hz, filter design: Butterworth IIR, filter order: 4
+        # 5. Integration of vertical speed --> vertical displacement d(t)
+        # 6. Compute total vertical displacement during the step (d_step):
+        # d_step = |max(d(t)) - min(d(t))|
+        # 7. Biomechanical model:
+        # StepLength = A * 2 * sqrt(2 * LBh * d_step - d_step^2)
+        # A: tuning coefficient, optimized by training for each population
+        # LBh: sensor height in meters, representative of the height of the center of mass
+        # TODO: Which is the correct way to pass tuning_coefficient, initial_contacts, sensor_height?
+        initial_contacts = np.squeeze(initial_contacts.astype(int))
+        sl_zjilstra_v3 = _zjilsV3(vacc_high, sampling_rate_hz, self.tuning_coefficient, initial_contacts, sensor_height)
+
+        # 8. Remove step length outliers during the gait sequence --> Hampel filter based on median absolute deviation
+        # 9. Interpolation of StepLength values --> Step length per second values
+        # 10. Remove step length per second outliers during the gait sequence --> Hampel filter based on median absolute deviation
+        duration = int(np.floor(len(vacc) / sampling_rate_hz))  # bottom-rounded duration (s)
+        ICtime = initial_contacts / sampling_rate_hz  # initial contacts: samples -> sec
+        sec_centers = (
+            np.arange(0, duration) + 1
+        )  # TODO: Differently from how it is done for cadence, here we consider second bins going from 0.5 to 1.5, from 1.5 to 2.5, etc...
+        # padding last value of the calculated step length values so that sl_zjilstra_v3 is as long as ICtime
+        if len(sl_zjilstra_v3) < len(ICtime):
+            sl_zjilstra_v3 = np.concatenate(
+                [sl_zjilstra_v3, np.tile(sl_zjilstra_v3[-1], len(ICtime) - len(sl_zjilstra_v3))]
+            )
+
+        slSec_zjilstra_v3 = robust_step_para_to_sec(
+            ICtime, sl_zjilstra_v3, sec_centers, self.max_interpolation_gap_s, self.step_length_smoothing.clone()
+        )
+        # 11. Approximated stride length per second values = 2* step length per second values
+        slSec = slSec_zjilstra_v3[0:duration]*2
+
+        # Results
+        self.sl_sec_list_ = pd.DataFrame(
+            {"stride_length_m": slSec}, index=as_samples(sec_centers, sampling_rate_hz)
+        ).rename_axis(index="sec_center_samples")
+        self.stpl_list_ = sl_zjilstra_v3
```

Comment:
Mmh `stpl_list_` might be hard to understand. Maybe we can use a better name. Might be the best to be more vorbose in both cases. So stride_length_per_sec... and step_length_per_sec... to avoid confusion

Reply:
Ah and we might want to have 3 outputs: Raw zjilstra output, interpolated zjilstra, interpolated and converted to strides.

Reply:
Seems legit. I renamed the three outputs, now it should be clearer:
- step_length_list_
- step_length_per_sec_list_
- stride_length_per_sec_list_